<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.2.6" />
<style type="text/css">
/* Debug borders */
p, li, dt, dd, div, pre, h1, h2, h3, h4, h5, h6 {
/*
  border: 1px solid red;
*/
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

tt {
  color: navy;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  font-family: sans-serif;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}

div.sectionbody {
  font-family: serif;
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}

pre {
  padding: 0;
  margin: 0;
}

span#author {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  font-size: 1.1em;
}
span#email {
}
span#revision {
  font-family: sans-serif;
}

div#footer {
  font-family: sans-serif;
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
div#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
div#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

div#preamble,
div.tableblock, div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-right: 10%;
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.5em;
  margin-bottom: 2.5em;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}

div.listingblock {
  margin-right: 0%;
}
div.listingblock > div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock > div.content {
  padding-left: 2.0em;
}

div.attribution {
  text-align: right;
}
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 2px solid silver;
}

div.exampleblock > div.content {
  border-left: 2px solid silver;
  padding: 0.5em;
}

div.verseblock div.content {
  white-space: pre;
}

div.imageblock div.content { padding-left: 0; }
div.imageblock img { border: 1px solid silver; }
span.image img { border-style: none; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
div.olist > ol {
  list-style-type: decimal;
}
div.olist2 > ol {
  list-style-type: lower-alpha;
}

div.tableblock > table {
  border: 3px solid #527bbd;
}
thead {
  font-family: sans-serif;
  font-weight: bold;
}
tfoot {
  font-weight: bold;
}

div.hlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hlist td {
  padding-bottom: 15px;
}
td.hlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
}
td.hlist2 {
  vertical-align: top;
}

@media print {
  div#footer-badges { display: none; }
}

div#toctitle {
  color: #527bbd;
  font-family: sans-serif;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}
/* Workarounds for IE6's broken and incomplete CSS2. */

div.sidebar-content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}
div.sidebar-title, div.image-title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  margin-top: 0.0em;
  margin-bottom: 0.5em;
}

div.listingblock div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock-content {
  padding-left: 2.0em;
}

div.exampleblock-content {
  border-left: 2px solid silver;
  padding-left: 0.5em;
}

/* IE6 sets dynamically generated links as visited. */
div#toc a:visited { color: blue; }

/* Because IE6 child selector is broken. */
div.olist2 ol {
  list-style-type: lower-alpha;
}
div.olist2 div.olist ol {
  list-style-type: decimal;
}
</style>
<title>VCFannotator: describing SNPs in genes and genomes</title>
</head>
<body>
<div id="header">
<h1>VCFannotator: describing SNPs in genes and genomes</h1>
</div>
<div id="preamble">
<div class="sectionbody">
<div class="para"><p>VCFannotator examines the positions of SNPs found within and between genes in an annotated genome sequence. SNPs localized within genes are annotated according to the gene structure, indicating whether the SNP is found within an untranslated region (UTR), intron, or coding exon. If the SNP is in a coding region, the impact of the SNP on the translated protein sequence is described (ex. synonomous or non-synonomous change).</p></div>
<div class="para"><p>The VCFannotator software was written at the <a href="http://www.broadinstitute.org">Broad Institute</a> to assist microbial genome analyses.</p></div>
</div>
</div>
<h2 id="_obtaining_vcfannotator">Obtaining VCFannotator</h2>
<div class="sectionbody">
<div class="para"><p>VCFannotator can be downloaded from <a href="https://sourceforge.net/projects/vcfannotator/files/">here</a>.</p></div>
</div>
<h2 id="_running_vcfannotator">Running VCFannotator</h2>
<div class="sectionbody">
<div class="para"><p>VCFannotator requires three input files: a genome sequence in FASTA format, gene annotations described in <a href="http://www.sequenceontology.org/gff3.shtml">GFF3</a> file format, and SNPs defined in <a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF</a> format.  Usage information is described below:</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>VCF_annotator.pl</tt></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><tt>##############################################################################
#
#  --gff3     gene annotations in gff3 format
#  --genome   genome sequence in fasta format.
#  --vcf      SNP data in vcf format
#
#  -X         write the .coding_mutations_described.txt file
#
##############################################################################</tt></pre>
</div></div>
</div>
<h2 id="_sample_data">Sample Data</h2>
<div class="sectionbody">
<div class="para"><p>Sample input data is provided in the <em>sample_data/</em> directory of the downloaded software. A <em>runMe.sh</em> script is provided to demonstrate software execution.</p></div>
</div>
<h2 id="_output_format">Output Format</h2>
<div class="sectionbody">
<div class="para"><p>The inputed VCF file is decorated with annotations by adding tab-delimited fields to each line of the VCF file.  If the SNP is localized to multiple overlapping genes or transcript isoforms, an annotation will be added separately for each feature.</p></div>
<div class="para"><p>An example output line transposed to a column format would look like so (taken from the sample data):</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>0   TY-2482_chromosome
1   5080
2   .
3   T
4   C
5   8101.55
6   PASS
7   AC=2;AF=1.00;AN=2;DP=212;Dels=0.00;FS=0.000;HRun=1;HaplotypeScore=0.0000;MQ=59.52;MQ0=0;QD=38.21;SB=-4130.96
8   GT:AD:DP:GQ:PL
9   1/1:0,212:212:99:8101,605,0
10  CDS,tmpTY2482_00008,tmpTY2482_00008T0,microcin H47 immunity protein mchI,trans_orient:+,loc_in_cds:46,codon_pos:1,codon:Tct-Cct,pep:S-&gt;P,Ser-16-Pro,(NSY)</tt></pre>
</div></div>
<div class="para"><p>The final field is bundled with the individual feature annotation data, comma-delimited. This includes the feature type (eg. CDS, intron, UTR), gene and transcript identifiers, name of the gene, and the transcribed orientation of the gene.  If the SNP is localized to a coding region, then the relative position within that CDS sequence is provided in addition to the codon change.  The types of coding mutations are provided as synonomous (SYN), non-synonomous (NSY), read-thru (RTH), and nonsence (STP).  SNPs that are localized to intergenic regions are reported as such along with the identifiers of the neighboring genes and distance to each.</p></div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2012-09-18 20:56:42 EDT
</div>
</div>
</body>
</html>
